File: golang/photolab/client_photolab.go
package photolab

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"mime/multipart"
	"net/http"
	"net/url"
	"os"
	"strings"
)

var (
	API_ENDPOINT        = "http://api-soft.photolab.me"
	API_UPLOAD_ENDPOINT = "http://upload-soft.photolab.me/upload.php"
	API_ENDPOINT_PROXY  = "http://api-proxy-soft.photolab.me"
)

type Steps struct {
	Steps []Step `json:"steps"`
}

type Step struct {
	Id        int64    `json:"id"`
	ImageUrls []string `json:"image_urls"`
}

type Image struct {
	Url      string  `xml:",innerxml"`
	Order    int     `xml:"order,attr"`
	Rectf    string  `xml:"rectf,attr,omitempty"`
	Rotation float32 `xml:"rotation,attr,omitempty"`
	Flip     int     `xml:"flip,attr,omitempty"`
}

type ImageRequest struct {
	Url    string
	Rotate int
	Flip   int
	Crop   string
}

type PhotolabClient struct {
}

func (p *PhotolabClient) PhotolabStepsAdvanced(comboId int64) (steps Steps, err error) {
	formData := url.Values{}
	formData.Set("combo_id", fmt.Sprint(comboId))

	endpoint := fmt.Sprintf("%s/photolab_steps_advanced.php", API_ENDPOINT)

	data, err := p.query(endpoint, formData)
	if err != nil {
		return
	}
	if err = json.Unmarshal(data, &steps); err != nil {
		return
	}
	return
}

func (p *PhotolabClient) TemplateProcess(templateName string, ims []ImageRequest) (s string, err error) {
	formData := url.Values{}
	formData.Set("template_name", templateName)
	for i, v := range ims {
		t := fmt.Sprintf("%d", i+1)
		formData.Set("template_name", templateName)
		formData.Set(fmt.Sprintf("image_url[%s]", t), v.Url)
		formData.Set(fmt.Sprintf("rotate[%s]", t), fmt.Sprint(v.Rotate))
		formData.Set(fmt.Sprintf("crop[%s]", t), v.Crop)
		formData.Set(fmt.Sprintf("flip[%s]", t), fmt.Sprint(v.Flip))
	}
	endpoint := fmt.Sprintf("%s/template_process.php", API_ENDPOINT)
	data, err := p.query(endpoint, formData)
	s = string(data)
	return
}

func (p *PhotolabClient) PhotolabProcess(templateName string, ims []ImageRequest) (s string, err error) {
	formData := url.Values{}
	formData.Set("template_name", templateName)
	for i, v := range ims {
		t := fmt.Sprintf("%d", i+1)
		formData.Set("template_name", templateName)
		formData.Set(fmt.Sprintf("image_url[%s]", t), v.Url)
		formData.Set(fmt.Sprintf("rotate[%s]", t), fmt.Sprint(v.Rotate))
		formData.Set(fmt.Sprintf("crop[%s]", t), v.Crop)
		formData.Set(fmt.Sprintf("flip[%s]", t), fmt.Sprint(v.Flip))
	}
	endpoint := fmt.Sprintf("%s/template_process.php", API_ENDPOINT)
	data, err := p.query(endpoint, formData)
	s = string(data)
	return
}

func (p *PhotolabClient) ImageUpload(image string) (s string, err error) {
	file, err := os.Open(image)
	if err != nil {
		return
	}
	defer file.Close()
	var requestBody bytes.Buffer
	multiPartWriter := multipart.NewWriter(&requestBody)
	fileWriter, err := multiPartWriter.CreateFormFile("file1", "image.jpg")
	if err != nil {
		return
	}
	_, err = io.Copy(fileWriter, file)
	if err != nil {
		return
	}
	fieldWriter, err := multiPartWriter.CreateFormField("no_resize")
	if err != nil {
		return
	}
	_, err = fieldWriter.Write([]byte("1"))
	if err != nil {
		return
	}
	multiPartWriter.Close()

	req, err := http.NewRequest("POST", API_UPLOAD_ENDPOINT, &requestBody)
	if err != nil {
		return
	}
	req.Header.Set("Content-Type", multiPartWriter.FormDataContentType())
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return
	}
	defer resp.Body.Close()
	data, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return
	}
	if resp.StatusCode != http.StatusOK {
		msg := fmt.Sprintf("StatusCode %d from: %s", resp.StatusCode, req.URL.String())
		msg += " Response:" + string(data)
		err = fmt.Errorf(msg)
		return
	}
	s = string(data)
	return
}

func (p *PhotolabClient) TemplateUpload(resources string) (s string, err error) {
	file, err := os.Open(resources)
	if err != nil {
		return
	}
	defer file.Close()
	var requestBody bytes.Buffer
	multiPartWriter := multipart.NewWriter(&requestBody)
	fileWriter, err := multiPartWriter.CreateFormFile("resources", "resources.zip")
	if err != nil {
		return
	}
	_, err = io.Copy(fileWriter, file)
	if err != nil {
		return
	}
	multiPartWriter.Close()
	endpoint := fmt.Sprintf("%s/template_upload.php", API_ENDPOINT_PROXY)
	req, err := http.NewRequest("POST", endpoint, &requestBody)
	if err != nil {
		return
	}
	req.Header.Set("Content-Type", multiPartWriter.FormDataContentType())
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return
	}
	defer resp.Body.Close()
	data, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return
	}
	if resp.StatusCode != http.StatusOK {
		msg := fmt.Sprintf("StatusCode %d from: %s", resp.StatusCode, req.URL.String())
		msg += " Response:" + string(data)
		err = fmt.Errorf(msg)
		return
	}
	s = string(data)
	return
}

func (p *PhotolabClient) DowloadFile(endpoint, dst string) (err error) {
	resp, err := http.Get(endpoint)
	if err != nil {
		return
	}
	defer resp.Body.Close()
	data, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return
	}
	if resp.StatusCode != http.StatusOK {
		msg := fmt.Sprintf("StatusCode %d from: %s", resp.StatusCode, endpoint)
		msg += " Response:" + string(data)
		err = fmt.Errorf(msg)
		return
	}
	err = ioutil.WriteFile(dst, data, 0666)
	return
}

func (p *PhotolabClient) query(endpoint string, formValues url.Values) (data []byte, err error) {
	httpReq, err := http.NewRequest("POST", endpoint, strings.NewReader(formValues.Encode()))
	if err != nil {
		return
	}
	httpReq.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	resp, err := http.DefaultClient.Do(httpReq)
	if err != nil {
		return
	}
	defer resp.Body.Close()
	data, err = ioutil.ReadAll(resp.Body)
	if err != nil {
		return
	}
	if resp.StatusCode != http.StatusOK {
		msg := fmt.Sprintf("StatusCode %d from: %s", resp.StatusCode, httpReq.URL.String())
		msg += " Response:" + string(data)
		err = fmt.Errorf(msg)
		return
	}
	return
}

File: java/src/main/java/me/photolab/ClientPhotolab.java
package me.photolab;

import com.google.gson.Gson;
// import okhttp3.*;
import okhttp3.OkHttpClient;
import okhttp3.FormBody;
import okhttp3.Request;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import okhttp3.Response;
import okhttp3.MediaType;

import okio.Buffer;
import okio.BufferedSink;
import okio.BufferedSource;
import okio.Okio;

import java.io.File;
import java.io.FileOutputStream;

class Steps {
    Step[] steps;
}

class Step {
    long id;
    String[] image_urls;
}

class ImageRequest {
    String url;
    int rotate;
    int flip;
    String crop;
}


public class ClientPhotolab {
    public static final String API_ENDPOINT = "http://api-soft.photolab.me";
    public static final String API_UPLOAD_ENDPOINT = "http://upload-soft.photolab.me/upload.php";
    public static final String API_ENDPOINT_PROXY = "http://api-proxy-soft.photolab.me";


    public Steps PhotolabStepsAdvanced(long comboId) throws Exception {
        RequestBody formBody = new FormBody.Builder()
                .add("combo_id", String.format("%d", comboId))
                .build();
        String endpoint = String.format("%s/photolab_steps_advanced.php", API_ENDPOINT);

        String data = this.query(endpoint, formBody);
        Gson gson = new Gson();
        Steps st = gson.fromJson(data, Steps.class);
        return st;
    }

    public String TemplateProcess(String template_name, ImageRequest[] ims) throws Exception {
        FormBody.Builder formBuilder = new FormBody.Builder();
        formBuilder.add("template_name", template_name);
        for (int i = 0; i < ims.length; i++) {
            ImageRequest im = ims[i];
            String t = String.format("%d", i + 1);
            formBuilder.add(String.format("image_url[%s]", t), im.url);
            formBuilder.add(String.format("rotate[%s]", t), Integer.toString(im.rotate));
            formBuilder.add(String.format("crop[%s]", t), im.crop);
            formBuilder.add(String.format("flip[%s]", t), Integer.toString(im.flip));
        }
        String endpoint = String.format("%s/template_process.php", API_ENDPOINT);
        Gson gson = new Gson();
        return this.query(endpoint, formBuilder.build());
    }

    public String PhotolabProcess(String template_name, ImageRequest[] ims) throws Exception {
        return TemplateProcess(template_name, ims);
    }

    public void DownloadFile(String url, String dest) throws Exception {
        OkHttpClient client = new OkHttpClient();
        Request request = new Request.Builder().url(url).build();
        Response response = client.newCall(request).execute();
        ResponseBody body = response.body();
        BufferedSource source = body.source();
        FileOutputStream destFile = new FileOutputStream(dest);
        BufferedSink sink = Okio.buffer(Okio.sink(destFile));
        Buffer sinkBuffer = sink.buffer();
        int bufferSize = 64 * 1024;
        for (; (source.read(sinkBuffer, bufferSize)) != -1; ) {
            sink.emit();
        }
        sink.flush();
        sink.close();
        source.close();
        return;
    }

    protected String query(String url, RequestBody formBody) throws Exception {
        OkHttpClient client = new OkHttpClient();
        Request request = new Request.Builder()
                .url(url)
                .post(formBody)
                .build();
        Response response = client.newCall(request).execute();
        if (!response.isSuccessful()) {
            String message = String.format("StatusCode %d from: %s", response.code(), request.url()) + " Response: " + response.body().string();
            throw new Exception(message);
        }
        return response.body().string();
    }

    public String ImageUpload(String image) throws Exception {
        String url = API_UPLOAD_ENDPOINT;
        File file = new File(image);
        RequestBody formBody = new MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("file1", file.getName(),
                        RequestBody.create(MediaType.parse("multipart/form-data"), file))
                .addFormDataPart("no_resize", "1")
                .build();

        return query(url, formBody);
    }

    public String TemplateUpload(String resources) throws Exception {
        String url = String.format("%s/template_upload.php", API_ENDPOINT_PROXY);
        File file = new File(resources);
        RequestBody formBody = new MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("resources", file.getName(),
                        RequestBody.create(MediaType.parse("multipart/form-data"), file))
                .build();
        return query(url, formBody);
    }
}
